istioctl/pkg/precheck/precheck_test.go

```diff
@@ -0,0 +1,166 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package precheck
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"istio.io/istio/istioctl/pkg/cli"
+	"istio.io/istio/pkg/config/analysis/diag"
+	"istio.io/istio/pkg/config/analysis/msg"
+	"istio.io/istio/pkg/kube"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/rest/fake"
+	cmdtesting "k8s.io/kubectl/pkg/cmd/testing"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+)
+
+type testCase struct {
+	version        string
+	revision       string
+	expectedOutput diag.Messages
+	expectedError  error
+}
+
+func Test_checkFromVersion(t *testing.T) {
+	cases := []testCase{
+		{
+			revision:       "canary",
+			version:        "1.21",
+			expectedOutput: diag.Messages{},
+			expectedError:  nil,
+		},
+		{
+			revision:       "canary",
+			version:        "XX",
+			expectedOutput: nil,
+			expectedError:  fmt.Errorf("invalid version XX, expected format like '1.0'"),
+		},
+		{
+			revision:       "canary",
+			version:        "2.0",
+			expectedOutput: nil,
+			expectedError:  fmt.Errorf("expected major version 1, got 2"),
+		},
+		{
+			revision:       "canary",
+			version:        "2.",
+			expectedOutput: nil,
+			expectedError:  fmt.Errorf("minor version is not a number: X"),
+		},
+		// TODO: add more test cases
+		// minor <= 21 and checkPilot failing
+		// minor <= 20 and checkDestinationRuleTLS failing
+		// minor <= 20 and checkExternalNameAlias failing
+		// minor <= 20 and checkVirtualServiceHostMatching failing
+		// minor <= 21 and checkPassthroughTargetPorts failing
+		// minor <= 21 and checkTracing failing
+	}
+
+	for _, c := range cases {
+		t.Run(fmt.Sprintf("revision=%s, version=%s", c.revision, c.version), func(t *testing.T) {
+			ctx := cli.NewFakeContext(&cli.NewFakeContextOption{
+				IstioNamespace: "istio-system",
+			})
+			output, err := checkFromVersion(ctx, c.revision, c.version)
+
+			assert.Equal(t, output, c.expectedOutput)
+
+			if c.expectedError != nil {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+func Test_checkTracing_ZipkinNotFound(t *testing.T) {
+	cli := kube.NewFakeClient()
+	messages := diag.Messages{}
+
+	// Test case: zipkin service not found
```

Comment:
I'm not sure we should error if a zipkin service is not found, based on the `checkTracing` function comment [here](https://github.com/istio/istio/blob/03df5acad6efae269471e5c245632cc567357ba1/istioctl/pkg/precheck/precheck.go#L183-L184):
```go
func checkTracing(cli kube.CLIClient, messages *diag.Messages) error {
	// In 1.22, we remove the default tracing config which points to zipkin.istio-system
	// This has no effect for users, unless they have this service.
```

IIUC, `checkTracing`, as it relates to `zipkin`, should not effect users unless they have this service.

Reply:
Do you mean that this should be removed: https://github.com/istio/istio/blob/03df5acad6efae269471e5c245632cc567357ba1/istioctl/pkg/precheck/precheck.go#L186-L188 ?

Reply:
I don't think so. It seems like the code already does what the comment says. I think your response below is the correct interpretation. Nil in the case we don't find zipkin when it's not expected to be found (which I think is fine because it's no longer a default). Should only be an issue if zipkin exists and is expected to exist but for some reason it comes back as not found, then we should error.